sycl/plugins/unified_runtime/ur/adapters/hip/context.hpp

```diff
@@ -100,9 +102,47 @@ struct ur_context_handle_t_ {
 
   uint32_t getReferenceCount() const noexcept { return RefCount; }
 
+  /// We need to keep track of USM mappings in AMD HIP, as certain extra
+  /// synchronization *is* actually required for correctness.
+  /// During kernel enqueue we must dispatch a prefetch for each kernel argument
+  /// that points to a USM mapping to ensure the mapping is correctly
+  /// populated on the device (https://github.com/intel/llvm/issues/7252). Thus,
+  /// we keep track of mappings in the context, and then check against them just
+  /// before the kernel is launched. The stream against which the kernel is
+  /// launched is not known until enqueue time, but the USM mappings can happen
+  /// at any time. Thus, they are tracked on the context used for the urUSM*
+  /// mapping.
+  ///
+  /// The three utility function are simple wrappers around a mapping from a
+  /// pointer to a size.
+  void addUSMMapping(void *Ptr, size_t Size) {
+    std::lock_guard<std::mutex> guard(Mutex);
+    assert(USMMappings.find(Ptr) == USMMappings.end() &&
+           "mapping already exists");
+    USMMappings[Ptr] = Size;
+  }
+
+  void removeUSMMapping(const void *ptr) {
+    std::lock_guard<std::mutex> guard(Mutex);
+    auto It = std::find_if(
+        USMMappings.begin(), USMMappings.end(),
+        [ptr](std::pair<const void *, size_t> val) { return val.first = ptr; });
```

Comment:
I guess this should be:
```suggestion
        [Ptr](std::pair<const void *, size_t> Val) { return Val.first == Ptr; });
```